swarm/storage/pusher/pusher_test.go

```diff
@@ -0,0 +1,259 @@
+package pusher
+
+import (
+	"context"
+	"encoding/binary"
+	"flag"
+	"fmt"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/swarm/chunk"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+	colorable "github.com/mattn/go-colorable"
+)
+
+var (
+	loglevel = flag.Int("loglevel", 3, "verbosity of logs")
+)
+
+func init() {
+	flag.Parse()
+	log.PrintOrigins(true)
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*loglevel), log.StreamHandler(colorable.NewColorableStderr(), log.TerminalFormat(true))))
+}
+
+// loopback implements PubSub as a central subscription engine,
+// ie a msg sent is received by all handlers registered for the topic
+type loopBack struct {
+	addr     []byte
+	handlers map[string][]func(msg []byte, p *p2p.Peer) error
+}
+
+func newLoopBack() *loopBack {
+	return &loopBack{
+		addr:     make([]byte, 32),
+		handlers: make(map[string][]func(msg []byte, p *p2p.Peer) error),
+	}
+}
+
+// Register subscribes to a topic with a handler
+func (lb *loopBack) Register(topic string, handler func(msg []byte, p *p2p.Peer) error) func() {
+	lb.handlers[topic] = append(lb.handlers[topic], handler)
+	return func() {}
+}
+
+// Send publishes a msg with a topic and directly calls registered handlers with
+// that topic
+func (lb *loopBack) Send(to []byte, topic string, msg []byte) error {
+	p := p2p.NewPeer(enode.ID{}, "", nil)
+	for _, handler := range lb.handlers[topic] {
+		if err := handler(msg, p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// BaseAddr needed to implement PubSub interface
+func (lb *loopBack) BaseAddr() []byte {
+	return lb.addr
+}
+
+// testPushSyncIndex mocks localstore and provides subscription and setting synced status
+// it implements PushSyncIndex interface
+type testPushSyncIndex struct {
+	i, total int       //
+	sent     *sync.Map // to store time of send for retry
+	synced   chan int  // to check if right amount of chunks
+}
+
+func newTestPushSyncIndex(chunkCnt int, sent *sync.Map) *testPushSyncIndex {
+	return &testPushSyncIndex{
+		i:      0,
+		total:  chunkCnt,
+		sent:   sent,
+		synced: make(chan int),
+	}
+}
+
+// SunscribePush allows iteration on the hashes
+func (t *testPushSyncIndex) SubscribePush(context.Context) (<-chan storage.Chunk, func()) {
```

Comment:
I find the comment here insufficient. It should also document the behavior when `total` is changed between calls (if I read it correctly, it will fill up with new chunks up to the new total).